compiler/aarch64/codegen/OMRLinkage.hpp

```diff
@@ -0,0 +1,439 @@
+
+/*
+ * The following #define and typedef must appear before any #includes in this file
+ */
+#ifndef OMR_LINKAGE_CONNECTOR
+#define OMR_LINKAGE_CONNECTOR
+namespace OMR { namespace ARM64 { class Linkage; } }
+namespace OMR { typedef OMR::ARM64::Linkage LinkageConnector; }
+#endif
+
+#include "compiler/codegen/OMRLinkage.hpp"
+
+#include <stddef.h>
+#include <stdint.h>
+#include "codegen/CodeGenerator.hpp"
+#include "codegen/InstOpCode.hpp"
+#include "codegen/RealRegister.hpp"
+#include "codegen/Register.hpp"
+#include "codegen/RegisterDependency.hpp"
+#include "env/TRMemory.hpp"
+
+class TR_FrontEnd;
+namespace TR { class AutomaticSymbol; }
+namespace TR { class Compilation; }
+namespace TR { class Instruction; }
+namespace TR { class MemoryReference; }
+namespace TR { class Node; }
+namespace TR { class ParameterSymbol; }
+namespace TR { class ResolvedMethodSymbol; }
+
+/**
+ * @brief Adds dependency
+ */
+inline void
+addDependency(
+      TR::RegisterDependencyConditions *dep,
+      TR::Register *vreg,
+      TR::RealRegister::RegNum rnum,
+      TR_RegisterKinds rk,
+      TR::CodeGenerator *cg)
+   {
+   if (vreg == NULL)
+      {
+      vreg = cg->allocateRegister(rk);
+      }
+
+   dep->addPreCondition(vreg, rnum);
+   dep->addPostCondition(vreg, rnum);
+   }
+
+namespace TR {
+
+class ARM64MemoryArgument
+   {
+   public:
+   //TR_ALLOC(TR_Memory::ARM64MemoryArgument)
+
+   TR::Register *argRegister;
+   TR::MemoryReference *argMemory;
+   TR::InstOpCode::Mnemonic opCode;
+   };
+
+
+// linkage properties
+#define CallerCleanup       0x01
+#define RightToLeft         0x02
+#define IntegersInRegisters 0x04
+#define LongsInRegisters    0x08
+#define FloatsInRegisters   0x10
+
+// register flags
+#define Preserved                   0x01
+#define IntegerReturn               0x02
+#define IntegerArgument             0x04
+#define FloatReturn                 0x08
+#define FloatArgument               0x10
+#define CallerAllocatesBackingStore 0x20
+#define ARM64_Reserved              0x40
+
+struct ARM64LinkageProperties
+   {
+   uint32_t _properties;
+   uint32_t _registerFlags[TR::RealRegister::NumRegisters];
+   uint8_t _numIntegerArgumentRegisters;
+   uint8_t _firstIntegerArgumentRegister;
+   uint8_t _numFloatArgumentRegisters;
+   uint8_t _firstFloatArgumentRegister;
+   TR::RealRegister::RegNum _argumentRegisters[TR::RealRegister::NumRegisters];
+   uint8_t _firstIntegerReturnRegister;
+   uint8_t _firstFloatReturnRegister;
+   TR::RealRegister::RegNum _returnRegisters[TR::RealRegister::NumRegisters];
+   uint8_t _numAllocatableIntegerRegisters;
+   uint8_t _numAllocatableFloatRegisters;
+   TR::RealRegister::RegNum _methodMetaDataRegister;
+   TR::RealRegister::RegNum _stackPointerRegister;
+   TR::RealRegister::RegNum _framePointerRegister;
+   TR::RealRegister::RegNum _vtableIndexArgumentRegister; // for icallVMprJavaSendPatchupVirtual
+   TR::RealRegister::RegNum _j9methodArgumentRegister;    // for icallVMprJavaSendStatic
```

Comment:
Let's not add these two registers for now.  They're OpenJ9-specific, and in the ARM backend (not AArch64) their only use is guarded by `J9_PROJECT_SPECIFIC` macros.  We should aspire to a better design in AArch64 if/when the time comes.